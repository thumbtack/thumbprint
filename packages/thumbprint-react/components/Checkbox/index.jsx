import React from 'react';
import PropTypes from 'prop-types';
import * as tokens from '@thumbtack/thumbprint-tokens';
import warning from 'warning';
import styles from './index.module.scss';

const labelCursor = {
    error: 'pointer',
    disabled: 'default',
    default: 'pointer',
};

const color = {
    error: {
        unchecked: tokens.tpColorRed,
        checked: tokens.tpColorRed,
        indeterminate: tokens.tpColorRed,
    },
    disabled: {
        unchecked: tokens.tpColorGray,
        checked: tokens.tpColorGray,
        indeterminate: tokens.tpColorGray,
    },
    default: {
        unchecked: 'inherit',
        checked: tokens.tpColorWhite,
        indeterminate: tokens.tpColorWhite,
    },
};

const borderColor = {
    error: {
        unchecked: tokens.tpColorRed,
        checked: tokens.tpColorRed,
        indeterminate: tokens.tpColorRed,
    },
    disabled: {
        unchecked: tokens.tpColorGray300,
        checked: tokens.tpColorGray300,
        indeterminate: tokens.tpColorGray300,
    },
    default: {
        unchecked: tokens.tpColorGray,
        checked: tokens.tpColorBlue,
        indeterminate: tokens.tpColorBlue,
    },
};

const backgroundColor = {
    error: {
        unchecked: tokens.tpColorWhite,
        checked: tokens.tpColorWhite,
        indeterminate: tokens.tpColorWhite,
    },
    disabled: {
        unchecked: tokens.tpColorGray200,
        checked: tokens.tpColorGray200,
        indeterminate: tokens.tpColorGray200,
    },
    default: {
        unchecked: tokens.tpColorWhite,
        checked: tokens.tpColorBlue,
        indeterminate: tokens.tpColorBlue,
    },
};

const textColor = {
    error: tokens.tpColorRed,
    disabled: tokens.tpColorGray,
    default: 'inherit',
};

const getCheckedState = ({ isChecked, isIndeterminate }) => {
    if (isChecked) {
        return 'checked';
    }

    if (isIndeterminate) {
        return 'indeterminate';
    }

    return 'unchecked';
};

const getFunctionalState = ({ isDisabled, hasError }) => {
    if (isDisabled) {
        return 'disabled';
    }

    if (hasError) {
        return 'error';
    }

    return 'default';
};

const Checkbox = ({
    children,
    dataTest,
    hasError,
    id,
    isChecked,
    isDisabled,
    isIndeterminate,
    labelPadding,
    name,
    onChange,
}) => {
    const functionalState = getFunctionalState({ isDisabled, hasError });
    const checkedState = getCheckedState({ isChecked, isIndeterminate });

    warning(
        children || id,
        'If `children` is not provided to `Checkbox`, the `id` prop must be used to link the `Checkbox` to a `label` element.',
    );

    return (
        // eslint-disable-next-line jsx-a11y/label-has-associated-control, jsx-a11y/label-has-for
        <label
            className={styles.root}
            style={{ cursor: labelCursor[functionalState], padding: labelPadding }}
        >
            <input
                className={styles.input}
                aria-checked={isIndeterminate ? 'mixed' : isChecked}
                data-test={dataTest}
                type="checkbox"
                id={id}
                name={name}
                checked={isChecked}
                onChange={e => onChange(e.target.checked, id)}
                disabled={isDisabled}
            />

            <div
                className={styles.checkboxImage}
                style={{
                    color: color[functionalState][checkedState],
                    backgroundColor: backgroundColor[functionalState][checkedState],
                    borderColor: borderColor[functionalState][checkedState],
                }}
            >
                {isChecked && !isIndeterminate && (
                    <svg
                        xmlns="http://www.w3.org/2000/svg"
                        fill="currentColor"
                        width="18"
                        height="18"
                        viewBox="0 0 18 18"
                        data-thumbprint-id="contentactions-check--small"
                    >
                        <path
                            fillRule="evenodd"
                            d="M14.5810763,3.19812449 C14.132085,2.85681117 13.5060971,2.96739243 13.1851033,3.44586905 L8.01720309,11.1408364 L4.52827048,8.83245251 C4.05827956,8.52197434 3.44229146,8.67402358 3.1502971,9.17270255 C2.85930272,9.67138151 3.00329993,10.3274261 3.47229088,10.6379043 L8.55419272,14 L14.8130718,4.68246529 C15.1350656,4.20398867 15.0300676,3.53943781 14.5810763,3.19812449"
                        />
                    </svg>
                )}
                {isIndeterminate && (
                    <svg
                        width="10"
                        height="2"
                        viewBox="0 0 10 2"
                        fill="currentColor"
                        xmlns="http://www.w3.org/2000/svg"
                    >
                        <rect x="0" y="0" width="10" height="2" />
                    </svg>
                )}
            </div>

            {children && (
                <span style={{ color: textColor[functionalState] }} className={styles.text}>
                    {children}
                </span>
            )}
        </label>
    );
};

Checkbox.propTypes = {
    /**
     * Disables the input and the label.
     */
    isDisabled: PropTypes.bool,
    /**
     * Determines if the checkbox is checked.
     */
    isChecked: PropTypes.bool,
    /**
     * Makes the radio and text color red.
     */
    hasError: PropTypes.bool,
    /**
     * Text or elements that appear within the label. If `children` is not provided, the developer
     * must use the `Radio`'s `id` prop to associate it with a custom `<label>` element.
     */
    children: PropTypes.node,
    /**
     * The `id` is added to the checkbox as an HTML attribute and passed to the `onChange`
     * function.
     */
    id: PropTypes.string,
    /**
     * Checkboxes on a page with the same name will be grouped together when sent to the server.
     * The browser will only send the value of checkboxes that are checked.
     */
    name: PropTypes.string,
    /**
     * Determine that padding that gets applied to the label. This can be used
     * to increase the label's click target. The value supplied should be a
     * string with a unit such as `8px` or `8px 16px`.
     */
    labelPadding: PropTypes.string,
    /**
     * Function that runs when a checkbox value changes. It receives the new boolean value and
     * the provided `id` as such: `props.onChange(e.target.checked, props.id)`.
     */
    onChange: PropTypes.func.isRequired,
    /**
     * Shows a horizontal line to represent an indeterminate input.
     */
    isIndeterminate: PropTypes.bool,
    /**
     * A selector hook into the React component for use in automated testing environments. It is
     * applied internally to the `<input />` element.
     */
    dataTest: PropTypes.string,
};

Checkbox.defaultProps = {
    isDisabled: false,
    children: null,
    id: null,
    isChecked: false,
    hasError: false,
    name: null,
    labelPadding: '14px 0',
    isIndeterminate: false,
    dataTest: undefined,
};

export default Checkbox;
