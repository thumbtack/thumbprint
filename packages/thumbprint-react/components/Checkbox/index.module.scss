@import '~@thumbtack/thumbprint-tokens/dist/scss/_index';

.root {
    position: relative;
    display: flex;

    &CheckboxVerticalAlignTop {
        align-items: flex-start;
    }

    &CheckboxVerticalAlignCenter {
        align-items: center;
    }
}

.input {
    position: absolute;
    opacity: 0;
    z-index: -1;
    width: 1px;
    height: 1px;

    &:focus ~ .checkboxImage {
        box-shadow: 0 0 0px 4px $tp-color__gray-300;
    }
}

.checkboxImage {
    width: 20px;
    height: 20px;
    border-radius: $tp-border-radius__base;
    border-width: 2px;
    border-style: solid;
    display: flex;
    align-items: center;
    justify-content: center;
    flex: none;

    // Margin that makes the input appear decently aligned against
    // various Thumbprint <Text> sizes
    .rootCheckboxVerticalAlignTop & {
        margin-top: 1px;
    }
}

.text {
    // This must take up the remaining available space for content that must be right-aligned.
    flex: 1 0 0%;
    // 'min-width' is needed to allow text within the checkbox to truncate.
    min-width: 0;
    font-size: $tp-font__body__1__size;
    font-weight: 400;
    padding-left: 8px;
}
