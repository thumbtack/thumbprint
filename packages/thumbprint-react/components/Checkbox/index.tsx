import React from 'react';
import classNames from 'classnames';
import * as tokens from '@thumbtack/thumbprint-tokens';
import warning from 'warning';
import styles from './index.module.scss';

const labelCursor = {
    error: 'pointer',
    disabled: 'default',
    default: 'pointer',
};

const color = {
    error: {
        unchecked: tokens.tpColorRed,
        checked: tokens.tpColorRed,
        indeterminate: tokens.tpColorRed,
    },
    disabled: {
        unchecked: tokens.tpColorGray,
        checked: tokens.tpColorGray,
        indeterminate: tokens.tpColorGray,
    },
    default: {
        unchecked: 'inherit',
        checked: tokens.tpColorWhite,
        indeterminate: tokens.tpColorWhite,
    },
};

const borderColor = {
    error: {
        unchecked: tokens.tpColorRed,
        checked: tokens.tpColorRed,
        indeterminate: tokens.tpColorRed,
    },
    disabled: {
        unchecked: tokens.tpColorGray300,
        checked: tokens.tpColorGray300,
        indeterminate: tokens.tpColorGray300,
    },
    default: {
        unchecked: tokens.tpColorGray,
        checked: tokens.tpColorBlue,
        indeterminate: tokens.tpColorBlue,
    },
};

const backgroundColor = {
    error: {
        unchecked: tokens.tpColorWhite,
        checked: tokens.tpColorWhite,
        indeterminate: tokens.tpColorWhite,
    },
    disabled: {
        unchecked: tokens.tpColorGray200,
        checked: tokens.tpColorGray200,
        indeterminate: tokens.tpColorGray200,
    },
    default: {
        unchecked: tokens.tpColorWhite,
        checked: tokens.tpColorBlue,
        indeterminate: tokens.tpColorBlue,
    },
};

const textColor = {
    error: tokens.tpColorRed,
    disabled: tokens.tpColorGray,
    default: 'inherit',
};

type CheckedState = 'checked' | 'indeterminate' | 'unchecked';
type FunctionalState = 'disabled' | 'error' | 'default';

const getCheckedState = ({
    isChecked,
    isIndeterminate,
}: Pick<PropTypes, 'isChecked' | 'isIndeterminate'>): CheckedState => {
    if (isChecked) {
        return 'checked';
    }

    if (isIndeterminate) {
        return 'indeterminate';
    }

    return 'unchecked';
};

const getFunctionalState = ({
    isDisabled,
    hasError,
}: Pick<PropTypes, 'isDisabled' | 'hasError'>): FunctionalState => {
    if (isDisabled) {
        return 'disabled';
    }

    if (hasError) {
        return 'error';
    }

    return 'default';
};

interface PropTypes {
    /**
     * Disables the input and the label.
     */
    isDisabled?: boolean;
    /**
     * Determines if the checkbox is checked.
     */
    isChecked?: boolean;
    /**
     * Makes the radio and text color red.
     */
    hasError?: boolean;
    /**
     * Text or elements that appear within the label. If `children` is not provided, the developer
     * must use the `Radio`'s `id` prop to associate it with a custom `<label>` element.
     */
    children?: React.ReactNode;
    /**
     * The `id` is added to the checkbox as an HTML attribute and passed to the `onChange`
     * function.
     */
    id?: string;
    /**
     * Adds the `required` HTML attribute.
     */
    isRequired?: boolean;
    /**
     * Checkboxes on a page with the same name will be grouped together when sent to the server.
     * The browser will only send the value of checkboxes that are checked.
     */
    name?: string;
    /**
     * Determine that padding that gets applied to the label. This can be used
     * to increase the label's click target. The value supplied should be a
     * string with a unit such as `8px` or `8px 16px`.
     */
    labelPadding?: string;
    /**
     * Function that runs when a checkbox value changes. It receives the new boolean value and
     * the provided `id` as such: `props.onChange(e.target.checked, props.id)`.
     */
    onChange: (value: boolean, id?: string) => void;
    /**
     * Shows a horizontal line to represent an indeterminate input.
     */
    isIndeterminate?: boolean;
    /**
     * A selector hook into the React component for use in automated testing environments. It is
     * applied internally to the `<input />` element.
     */
    dataTest?: string;
    /**
     * Determines how the checkbox input will be vertically aligned relative to `props.children`.
     */
    checkboxVerticalAlign?: 'top' | 'center';
    /**
     * Determines the value that will be submitted if the checkbox is checked. The default value is
     * `'on'`.
     */
    value?: string | string[] | number;
}

export default function Checkbox({
    checkboxVerticalAlign = 'center',
    children,
    dataTest,
    hasError = false,
    id,
    isChecked = false,
    isDisabled = false,
    isIndeterminate = false,
    labelPadding = '14px 0',
    isRequired = false,
    name,
    onChange,
    value,
}: PropTypes): JSX.Element {
    const functionalState = getFunctionalState({ isDisabled, hasError });
    const checkedState = getCheckedState({ isChecked, isIndeterminate });

    warning(
        children || id,
        'If `children` is not provided to `Checkbox`, the `id` prop must be used to link the `Checkbox` to a `label` element.',
    );

    // React adds a `value` attribute (`value=""`) to `input[type="checkbox"]` even if the `value`
    // prop is `undefined`. This prevents the default browser behavior of `value="on"` when the
    // `value` attribute is omitted. We can work around the React behavior and avoid adding
    // `value=""` to the DOM by conditionally creating an object that we then spread onto the
    // element. More context: https://github.com/thumbtack/thumbprint/issues/589
    const valuePropObject = value ? { value } : {};

    return (
        // eslint-disable-next-line jsx-a11y/label-has-associated-control, jsx-a11y/label-has-for
        <label
            className={classNames(styles.root, {
                [styles.rootCheckboxVerticalAlignTop]: checkboxVerticalAlign === 'top',
                [styles.rootCheckboxVerticalAlignCenter]: checkboxVerticalAlign === 'center',
            })}
            style={{ cursor: labelCursor[functionalState], padding: labelPadding }}
        >
            <input
                className={styles.input}
                aria-checked={isIndeterminate ? 'mixed' : isChecked}
                data-test={dataTest}
                type="checkbox"
                id={id}
                name={name}
                checked={isChecked}
                onChange={(event): void => onChange(event.target.checked, id)}
                disabled={isDisabled}
                required={isRequired}
                {...valuePropObject}
            />

            <div
                className={styles.checkboxImage}
                style={{
                    color: color[functionalState][checkedState],
                    backgroundColor: backgroundColor[functionalState][checkedState],
                    borderColor: borderColor[functionalState][checkedState],
                }}
            >
                {isChecked && !isIndeterminate && (
                    <svg
                        width="18"
                        height="18"
                        viewBox="0 0 18 18"
                        fill="currentColor"
                        xmlns="http://www.w3.org/2000/svg"
                    >
                        <rect
                            x="15.232"
                            y="4.003"
                            width="11.701"
                            height="1.879"
                            rx=".939"
                            transform="rotate(123 15.232 4.003)"
                        />
                        <rect
                            x="8.83"
                            y="13.822"
                            width="7.337"
                            height="1.879"
                            rx=".939"
                            transform="rotate(-146 8.83 13.822)"
                        />
                        <path d="M8.072 13.306l1.03-1.586.787.512-1.03 1.586z" />
                    </svg>
                )}
                {isIndeterminate && (
                    <svg
                        width="10"
                        height="2"
                        viewBox="0 0 10 2"
                        fill="currentColor"
                        xmlns="http://www.w3.org/2000/svg"
                    >
                        <rect x="0" y="0" width="10" height="2" />
                    </svg>
                )}
            </div>

            {children && (
                <span style={{ color: textColor[functionalState] }} className={styles.text}>
                    {children}
                </span>
            )}
        </label>
    );
}
