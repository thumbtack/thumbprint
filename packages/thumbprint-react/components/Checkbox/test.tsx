import React from 'react';
import { mount } from 'enzyme';
import Checkbox from './index';

test('renders checkbox input', () => {
    const wrapper = mount(<Checkbox onChange={jest.fn()}>Goose</Checkbox>);
    expect(wrapper.find('input').prop('type')).toBe('checkbox');
    expect(wrapper).toMatchSnapshot();
});

test('renders `children` passed in', () => {
    const wrapper = mount(<Checkbox onChange={jest.fn()}>Goose</Checkbox>);
    expect(wrapper.text()).toBe('Goose');
    expect(wrapper).toMatchSnapshot();
});

test('if no `children` are passed in', () => {
    const wrapper = mount(<Checkbox onChange={jest.fn()} id="no-children" />);
    expect(wrapper).toMatchSnapshot();
});

test('renders `input` within `label`', () => {
    const wrapper = mount(<Checkbox onChange={jest.fn()}>Goose</Checkbox>);
    expect(wrapper.find('label input').exists()).toBe(true);
    expect(wrapper).toMatchSnapshot();
});

test('adds `disabled` attribute', () => {
    const wrapper = mount(
        <Checkbox isDisabled onChange={jest.fn()}>
            Goose
        </Checkbox>,
    );
    expect(wrapper.find('input').prop('disabled')).toBeTruthy();
    expect(wrapper).toMatchSnapshot();
});

test('adds `name` attribute', () => {
    const wrapper = mount(
        <Checkbox name="ohhi" onChange={jest.fn()}>
            Goose
        </Checkbox>,
    );
    expect(wrapper.find('input').prop('name')).toBe('ohhi');
    expect(wrapper).toMatchSnapshot();
});

test('is not checked by default', () => {
    const wrapper = mount(<Checkbox onChange={jest.fn()}>Goose</Checkbox>);
    expect(wrapper.find('input').prop('checked')).toBeFalsy();
    expect(wrapper).toMatchSnapshot();
});

test('is checked when `isChecked` is set to true', () => {
    const wrapper = mount(
        <Checkbox isChecked onChange={jest.fn()}>
            Goose
        </Checkbox>,
    );
    expect(wrapper.find('input').prop('checked')).toBe(true);
    expect(wrapper).toMatchSnapshot();
});

test('applies `labelPadding` as padding on the label', () => {
    const wrapper = mount(
        <Checkbox labelPadding="10px 20px" onChange={jest.fn()}>
            Goose
        </Checkbox>,
    );

    expect((wrapper.find('label').prop('style') || {}).padding).toBe('10px 20px');
    expect(wrapper).toMatchSnapshot();
});

test('applies className corresponding to `checkboxVerticalAlign` on the label', () => {
    const wrapper = mount(
        <Checkbox checkboxVerticalAlign="top" onChange={jest.fn()}>
            Goose
        </Checkbox>,
    );

    expect(wrapper.find('label').hasClass('rootCheckboxVerticalAlignTop'));
    expect(wrapper).toMatchSnapshot();
});

test('renders `hasError` state', () => {
    const wrapperA = mount(
        <Checkbox isChecked hasError onChange={jest.fn()}>
            Goose
        </Checkbox>,
    );
    const wrapperB = mount(
        <Checkbox hasError onChange={jest.fn()}>
            Goose
        </Checkbox>,
    );

    expect(wrapperA).toMatchSnapshot();
    expect(wrapperB).toMatchSnapshot();
});

test('calls `onChange` function when supplied', () => {
    const onChange = jest.fn();

    const wrapper = mount(<Checkbox onChange={onChange}>Goose</Checkbox>);

    wrapper.find('input').simulate('change', { target: { checked: true } });
    expect(onChange).toHaveBeenCalledTimes(1);

    wrapper.find('input').simulate('change', { target: { checked: true } });
    expect(onChange).toHaveBeenCalledTimes(2);
});

test('passes correct new value and `id` to `onChange` function', () => {
    const onChange = jest.fn();

    const wrapper = mount(
        <Checkbox onChange={onChange} id="goose" isChecked>
            Goose
        </Checkbox>,
    );

    wrapper.find('input').simulate('change', { target: { checked: false } });
    expect(onChange).toHaveBeenCalledWith(false, 'goose');

    wrapper.find('input').simulate('change', { target: { checked: true } });
    expect(onChange).toHaveBeenCalledWith(true, 'goose');
});

test('calls `onChange` with correct values when `isIndeterminate` is true ', () => {
    const onChange = jest.fn();
    const wrapper = mount(
        <Checkbox onChange={onChange} isIndeterminate isChecked>
            Goose
        </Checkbox>,
    );

    wrapper.find('input').simulate('change', { target: { checked: false } });
    expect(onChange).toHaveBeenCalledWith(false, undefined);

    wrapper.find('input').simulate('change', { target: { checked: true } });
    expect(onChange).toHaveBeenCalledWith(true, undefined);
});

test('renders indeterminate SVG when both `isIndeterminate` and `isChecked` are true ', () => {
    const wrapper = mount(
        <Checkbox isIndeterminate isChecked onChange={jest.fn()}>
            Goose
        </Checkbox>,
    );
    expect(wrapper.find('svg')).toHaveLength(1);
    expect(wrapper).toMatchSnapshot();
});

test('adds `dataTest` prop', () => {
    const wrapper = mount(
        <Checkbox dataTest="Duck" onChange={jest.fn()}>
            Goose
        </Checkbox>,
    );
    expect(wrapper.find('input').prop('data-test')).toEqual('Duck');
    expect(wrapper).toMatchSnapshot();
});

test('adds `value` prop', () => {
    const wrapper = mount(
        <Checkbox value="Duck" onChange={jest.fn()}>
            Goose
        </Checkbox>,
    );
    expect(wrapper.find('input').prop('value')).toEqual('Duck');
    expect(wrapper).toMatchSnapshot();
});
