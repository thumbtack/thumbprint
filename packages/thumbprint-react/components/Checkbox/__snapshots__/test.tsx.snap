// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`adds \`dataTest\` prop 1`] = `
<Checkbox
  dataTest="Duck"
  onChange={[MockFunction]}
>
  <label
    className="root rootCheckboxVerticalAlignCenter"
    style={
      Object {
        "cursor": "pointer",
        "padding": "14px 0",
      }
    }
  >
    <input
      aria-checked={false}
      checked={false}
      className="input"
      data-test="Duck"
      disabled={false}
      onChange={[Function]}
      required={false}
      type="checkbox"
    />
    <div
      className="checkboxImage"
      style={
        Object {
          "backgroundColor": "#ffffff",
          "borderColor": "#d3d4d5",
          "color": "inherit",
        }
      }
    />
    <span
      className="text"
      style={
        Object {
          "color": "inherit",
        }
      }
    >
      Goose
    </span>
  </label>
</Checkbox>
`;

exports[`adds \`disabled\` attribute 1`] = `
<Checkbox
  isDisabled={true}
  onChange={[MockFunction]}
>
  <label
    className="root rootCheckboxVerticalAlignCenter"
    style={
      Object {
        "cursor": "default",
        "padding": "14px 0",
      }
    }
  >
    <input
      aria-checked={false}
      checked={false}
      className="input"
      disabled={true}
      onChange={[Function]}
      required={false}
      type="checkbox"
    />
    <div
      className="checkboxImage"
      style={
        Object {
          "backgroundColor": "#fafafa",
          "borderColor": "#e9eced",
          "color": "#d3d4d5",
        }
      }
    />
    <span
      className="text"
      style={
        Object {
          "color": "#d3d4d5",
        }
      }
    >
      Goose
    </span>
  </label>
</Checkbox>
`;

exports[`adds \`name\` attribute 1`] = `
<Checkbox
  name="ohhi"
  onChange={[MockFunction]}
>
  <label
    className="root rootCheckboxVerticalAlignCenter"
    style={
      Object {
        "cursor": "pointer",
        "padding": "14px 0",
      }
    }
  >
    <input
      aria-checked={false}
      checked={false}
      className="input"
      disabled={false}
      name="ohhi"
      onChange={[Function]}
      required={false}
      type="checkbox"
    />
    <div
      className="checkboxImage"
      style={
        Object {
          "backgroundColor": "#ffffff",
          "borderColor": "#d3d4d5",
          "color": "inherit",
        }
      }
    />
    <span
      className="text"
      style={
        Object {
          "color": "inherit",
        }
      }
    >
      Goose
    </span>
  </label>
</Checkbox>
`;

exports[`adds \`value\` prop 1`] = `
<Checkbox
  onChange={[MockFunction]}
  value="Duck"
>
  <label
    className="root rootCheckboxVerticalAlignCenter"
    style={
      Object {
        "cursor": "pointer",
        "padding": "14px 0",
      }
    }
  >
    <input
      aria-checked={false}
      checked={false}
      className="input"
      disabled={false}
      onChange={[Function]}
      required={false}
      type="checkbox"
      value="Duck"
    />
    <div
      className="checkboxImage"
      style={
        Object {
          "backgroundColor": "#ffffff",
          "borderColor": "#d3d4d5",
          "color": "inherit",
        }
      }
    />
    <span
      className="text"
      style={
        Object {
          "color": "inherit",
        }
      }
    >
      Goose
    </span>
  </label>
</Checkbox>
`;

exports[`applies \`labelPadding\` as padding on the label 1`] = `
<Checkbox
  labelPadding="10px 20px"
  onChange={[MockFunction]}
>
  <label
    className="root rootCheckboxVerticalAlignCenter"
    style={
      Object {
        "cursor": "pointer",
        "padding": "10px 20px",
      }
    }
  >
    <input
      aria-checked={false}
      checked={false}
      className="input"
      disabled={false}
      onChange={[Function]}
      required={false}
      type="checkbox"
    />
    <div
      className="checkboxImage"
      style={
        Object {
          "backgroundColor": "#ffffff",
          "borderColor": "#d3d4d5",
          "color": "inherit",
        }
      }
    />
    <span
      className="text"
      style={
        Object {
          "color": "inherit",
        }
      }
    >
      Goose
    </span>
  </label>
</Checkbox>
`;

exports[`applies className corresponding to \`checkboxVerticalAlign\` on the label 1`] = `
<Checkbox
  checkboxVerticalAlign="top"
  onChange={[MockFunction]}
>
  <label
    className="root rootCheckboxVerticalAlignTop"
    style={
      Object {
        "cursor": "pointer",
        "padding": "14px 0",
      }
    }
  >
    <input
      aria-checked={false}
      checked={false}
      className="input"
      disabled={false}
      onChange={[Function]}
      required={false}
      type="checkbox"
    />
    <div
      className="checkboxImage"
      style={
        Object {
          "backgroundColor": "#ffffff",
          "borderColor": "#d3d4d5",
          "color": "inherit",
        }
      }
    />
    <span
      className="text"
      style={
        Object {
          "color": "inherit",
        }
      }
    >
      Goose
    </span>
  </label>
</Checkbox>
`;

exports[`if no \`children\` are passed in 1`] = `
<Checkbox
  id="no-children"
  onChange={[MockFunction]}
>
  <label
    className="root rootCheckboxVerticalAlignCenter"
    style={
      Object {
        "cursor": "pointer",
        "padding": "14px 0",
      }
    }
  >
    <input
      aria-checked={false}
      checked={false}
      className="input"
      disabled={false}
      id="no-children"
      onChange={[Function]}
      required={false}
      type="checkbox"
    />
    <div
      className="checkboxImage"
      style={
        Object {
          "backgroundColor": "#ffffff",
          "borderColor": "#d3d4d5",
          "color": "inherit",
        }
      }
    />
  </label>
</Checkbox>
`;

exports[`is checked when \`isChecked\` is set to true 1`] = `
<Checkbox
  isChecked={true}
  onChange={[MockFunction]}
>
  <label
    className="root rootCheckboxVerticalAlignCenter"
    style={
      Object {
        "cursor": "pointer",
        "padding": "14px 0",
      }
    }
  >
    <input
      aria-checked={true}
      checked={true}
      className="input"
      disabled={false}
      onChange={[Function]}
      required={false}
      type="checkbox"
    />
    <div
      className="checkboxImage"
      style={
        Object {
          "backgroundColor": "#009fd9",
          "borderColor": "#009fd9",
          "color": "#ffffff",
        }
      }
    >
      <svg
        fill="currentColor"
        height="18"
        viewBox="0 0 18 18"
        width="18"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          height="1.879"
          rx=".939"
          transform="rotate(123 15.232 4.003)"
          width="11.701"
          x="15.232"
          y="4.003"
        />
        <rect
          height="1.879"
          rx=".939"
          transform="rotate(-146 8.83 13.822)"
          width="7.337"
          x="8.83"
          y="13.822"
        />
        <path
          d="M8.072 13.306l1.03-1.586.787.512-1.03 1.586z"
        />
      </svg>
    </div>
    <span
      className="text"
      style={
        Object {
          "color": "inherit",
        }
      }
    >
      Goose
    </span>
  </label>
</Checkbox>
`;

exports[`is not checked by default 1`] = `
<Checkbox
  onChange={[MockFunction]}
>
  <label
    className="root rootCheckboxVerticalAlignCenter"
    style={
      Object {
        "cursor": "pointer",
        "padding": "14px 0",
      }
    }
  >
    <input
      aria-checked={false}
      checked={false}
      className="input"
      disabled={false}
      onChange={[Function]}
      required={false}
      type="checkbox"
    />
    <div
      className="checkboxImage"
      style={
        Object {
          "backgroundColor": "#ffffff",
          "borderColor": "#d3d4d5",
          "color": "inherit",
        }
      }
    />
    <span
      className="text"
      style={
        Object {
          "color": "inherit",
        }
      }
    >
      Goose
    </span>
  </label>
</Checkbox>
`;

exports[`renders \`children\` passed in 1`] = `
<Checkbox
  onChange={[MockFunction]}
>
  <label
    className="root rootCheckboxVerticalAlignCenter"
    style={
      Object {
        "cursor": "pointer",
        "padding": "14px 0",
      }
    }
  >
    <input
      aria-checked={false}
      checked={false}
      className="input"
      disabled={false}
      onChange={[Function]}
      required={false}
      type="checkbox"
    />
    <div
      className="checkboxImage"
      style={
        Object {
          "backgroundColor": "#ffffff",
          "borderColor": "#d3d4d5",
          "color": "inherit",
        }
      }
    />
    <span
      className="text"
      style={
        Object {
          "color": "inherit",
        }
      }
    >
      Goose
    </span>
  </label>
</Checkbox>
`;

exports[`renders \`hasError\` state 1`] = `
<Checkbox
  hasError={true}
  isChecked={true}
  onChange={[MockFunction]}
>
  <label
    className="root rootCheckboxVerticalAlignCenter"
    style={
      Object {
        "cursor": "pointer",
        "padding": "14px 0",
      }
    }
  >
    <input
      aria-checked={true}
      checked={true}
      className="input"
      disabled={false}
      onChange={[Function]}
      required={false}
      type="checkbox"
    />
    <div
      className="checkboxImage"
      style={
        Object {
          "backgroundColor": "#ffffff",
          "borderColor": "#ff5a5f",
          "color": "#ff5a5f",
        }
      }
    >
      <svg
        fill="currentColor"
        height="18"
        viewBox="0 0 18 18"
        width="18"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          height="1.879"
          rx=".939"
          transform="rotate(123 15.232 4.003)"
          width="11.701"
          x="15.232"
          y="4.003"
        />
        <rect
          height="1.879"
          rx=".939"
          transform="rotate(-146 8.83 13.822)"
          width="7.337"
          x="8.83"
          y="13.822"
        />
        <path
          d="M8.072 13.306l1.03-1.586.787.512-1.03 1.586z"
        />
      </svg>
    </div>
    <span
      className="text"
      style={
        Object {
          "color": "#ff5a5f",
        }
      }
    >
      Goose
    </span>
  </label>
</Checkbox>
`;

exports[`renders \`hasError\` state 2`] = `
<Checkbox
  hasError={true}
  onChange={[MockFunction]}
>
  <label
    className="root rootCheckboxVerticalAlignCenter"
    style={
      Object {
        "cursor": "pointer",
        "padding": "14px 0",
      }
    }
  >
    <input
      aria-checked={false}
      checked={false}
      className="input"
      disabled={false}
      onChange={[Function]}
      required={false}
      type="checkbox"
    />
    <div
      className="checkboxImage"
      style={
        Object {
          "backgroundColor": "#ffffff",
          "borderColor": "#ff5a5f",
          "color": "#ff5a5f",
        }
      }
    />
    <span
      className="text"
      style={
        Object {
          "color": "#ff5a5f",
        }
      }
    >
      Goose
    </span>
  </label>
</Checkbox>
`;

exports[`renders \`input\` within \`label\` 1`] = `
<Checkbox
  onChange={[MockFunction]}
>
  <label
    className="root rootCheckboxVerticalAlignCenter"
    style={
      Object {
        "cursor": "pointer",
        "padding": "14px 0",
      }
    }
  >
    <input
      aria-checked={false}
      checked={false}
      className="input"
      disabled={false}
      onChange={[Function]}
      required={false}
      type="checkbox"
    />
    <div
      className="checkboxImage"
      style={
        Object {
          "backgroundColor": "#ffffff",
          "borderColor": "#d3d4d5",
          "color": "inherit",
        }
      }
    />
    <span
      className="text"
      style={
        Object {
          "color": "inherit",
        }
      }
    >
      Goose
    </span>
  </label>
</Checkbox>
`;

exports[`renders checkbox input 1`] = `
<Checkbox
  onChange={[MockFunction]}
>
  <label
    className="root rootCheckboxVerticalAlignCenter"
    style={
      Object {
        "cursor": "pointer",
        "padding": "14px 0",
      }
    }
  >
    <input
      aria-checked={false}
      checked={false}
      className="input"
      disabled={false}
      onChange={[Function]}
      required={false}
      type="checkbox"
    />
    <div
      className="checkboxImage"
      style={
        Object {
          "backgroundColor": "#ffffff",
          "borderColor": "#d3d4d5",
          "color": "inherit",
        }
      }
    />
    <span
      className="text"
      style={
        Object {
          "color": "inherit",
        }
      }
    >
      Goose
    </span>
  </label>
</Checkbox>
`;

exports[`renders indeterminate SVG when both \`isIndeterminate\` and \`isChecked\` are true  1`] = `
<Checkbox
  isChecked={true}
  isIndeterminate={true}
  onChange={[MockFunction]}
>
  <label
    className="root rootCheckboxVerticalAlignCenter"
    style={
      Object {
        "cursor": "pointer",
        "padding": "14px 0",
      }
    }
  >
    <input
      aria-checked="mixed"
      checked={true}
      className="input"
      disabled={false}
      onChange={[Function]}
      required={false}
      type="checkbox"
    />
    <div
      className="checkboxImage"
      style={
        Object {
          "backgroundColor": "#009fd9",
          "borderColor": "#009fd9",
          "color": "#ffffff",
        }
      }
    >
      <svg
        fill="currentColor"
        height="2"
        viewBox="0 0 10 2"
        width="10"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          height="2"
          width="10"
          x="0"
          y="0"
        />
      </svg>
    </div>
    <span
      className="text"
      style={
        Object {
          "color": "inherit",
        }
      }
    >
      Goose
    </span>
  </label>
</Checkbox>
`;
